File: 0/0/3/003.hs
-- Copyright (C) 2013 Jorge Aparicio

import Data.Numbers.Primes (primes)

main :: IO()
main
  = print
  . last
  . primeFactors
  $ number
    where number = 600851475143 :: Int

primeFactors :: Integral a => a -> [a]
primeFactors n = factor n primes

factor :: Integral a => a -> [a] -> [a]
factor n (p:ps)
  | p*p > n        = [n]
  | n `mod` p == 0 = p : factor (n `div` p) (p:ps)
  | otherwise      = factor n ps

File: 0/0/3/003.c
// Copyright (C) 2013 Jorge Aparicio

#include <stdio.h>

int main() {
  long long n = 600851475143;
  int factor = 2;

  while (1) {
    if (n % factor == 0)
      n /= factor;
    else
      factor += 1;

    if (factor * factor > n) {
      factor = n;
      break;
    }

    if (n == 1)
      break;
  }

  printf("%d\n", factor);

  return 0;
}
